File: scala/src/main/scala/fun/pierre/lists/P02.scala
package fun.pierre.lists

import java.util.NoSuchElementException

import scala.annotation.tailrec

object P02 {

  @tailrec
  def lastAndPenultimateRecursion[T](list: List[T]): List[T] = list match {
    case x :: y :: Nil => x :: y :: Nil
    case _ :: x :: y => lastAndPenultimateRecursion(x :: y)
    case _ => throw new NoSuchElementException("list is empty")
  }
}

File: java/src/main/java/fun/pierre/lists/P02.java
package fun.pierre.lists;

import java.util.List;

public class P02 {
    public static<T> List<T> lastAndPenultimate(List<T> _list) {
        int size = _list.size();
        if (size <= 1) {
            throw new IllegalArgumentException("List has less than two elements");
        }

        return _list.subList(size - 2, size);
    }
}
